Bug24539

shalperin opened the pull request:
Added species list to search/csv export.  Refactor to ogr_conversion()..

shalperin committed:
Unit test for zip files

Added test that can check if a zip file's file
list is what is expected, and added this test to the
export feature.

shalperin committed:
refactored ogr_conversion

refactored ogr_conversion to take a variable list of named
sql statements.

shalperin committed:
removed old version of ogr_convert

shalperin committed:
bug 24539 add species list to export

added a call to ogr_convert that includes a species list in CSV mode.

ahinz commented on treemap/tests.py:
```diff
@@ -1289,25 +1293,50 @@ def test_need_permission_to_approve_pending(self):
-    def test_ogr(self):
+    def _check_zip_response_contains_files(self, response, files):
+        tmp_dir = tempfile.mkdtemp()
+        tmp_file = tmp_dir + '/attachment.zip'
```
We should use http://docs.python.org/library/os.path.html#os.path.join to join strings so that the tests will run on windows

ahinz commented on treemap/tests.py:
```diff
@@ -1289,25 +1293,50 @@ def test_need_permission_to_approve_pending(self):
+        f = open(tmp_file, 'w')
+        f.write(response.content)
+        f.close()
+        if (zipfile.is_zipfile(tmp_file) ==False):
```
Generally we're trying to stick to the python coding style where we don't use parens on if statements unless needed so a better formatting might be:

``` python
if not zipfile.is_zipfile(tmp_file):
     return ...
```

ahinz commented on treemap/tests.py:
```diff
@@ -1289,25 +1293,50 @@ def test_need_permission_to_approve_pending(self):
+        
+        zf = zipfile.ZipFile(tmp_file, 'r')
+        if (sorted(files) != sorted(zf.namelist()) ):
+            return "error: file list in %s is: %s but I expected: %s" % (tmp_file, sorted(zf.namelist()), sorted(files))
```
Instead of returning strings we should put asserts in here:

``` python
self.assertTrue(sorted(files) == sorted(zf.namelist(), msg="error: ...")
```

ahinz commented on treemap/tests.py:
```diff
@@ -1289,25 +1293,50 @@ def test_need_permission_to_approve_pending(self):
+        
+        zf.close()
+        shutil.rmtree(tmp_dir) 
+        return "OK"
```
This method should just return `true` if we get  to this point. Returning a specialized string (i.e. `"ok"`) is probably just asking for bugs later.

ahinz commented on treemap/tests.py:
```diff
@@ -1289,25 +1293,50 @@ def test_need_permission_to_approve_pending(self):
         self.assertEqual(response['content-disposition'], 'attachment; filename=trees.zip')
         self.assertNotEqual(len(response.content), 0)
-
+        self.assertEqual("OK", self._check_zip_response_contains_files(response, ["trees.kml", "plots.kml"]))   
```
Now all of these `self.assertEqual("OK", self._check_zip_response...)` can simply be:

``` python
self.assertTrue(...)
```

ahinz commented on treemap/views.py:
```diff
@@ -1815,11 +1815,8 @@ def zip_files(file_paths,archive_name):
         return zip_stream
 
-def ogr_conversion(output_type, tree_sql, plot_sql, extension=None, name="trees", geo=True):   
+def ogr_conversion(output_type, named_sql, extension=None, name="trees", geo=True):   
```
This method probably needs documentation that at least explains the expected format of `named_sql` and what the reset of the arguments do. This might also be a good place to explain why we use OGR and what need to be installed?

ahinz commented on treemap/views.py:
```diff
@@ -1827,48 +1824,47 @@ def ogr_conversion(output_type, tree_sql, plot_sql, extension=None, name="trees"
+        sql = s["sql"]
+        if "srs" in named_sql:
+            srs = s["srs"]
+        else: srs = 'EPSG:4326'
```
Although valid python we always put a newline after `else:` blocks

bhuber replied:
better yet: srs = s["srs"] if "srs" in named_sql else 'EPSG:4326' .  I used to hate the python inline if-else, but it's been growing on me.

Also, I have no idea what this is supposed to be doing so this could be totally wrong, but based on the surrounding code it looks like this should be 'if "srs" in s:', not 'if "srs" in named_sql:'.  If it is supposed to be named_sql, this statement can be moved outside the loop.

ahinz commented on treemap/views.py:
```diff
@@ -1827,48 +1824,47 @@ def ogr_conversion(output_type, tree_sql, plot_sql, extension=None, name="trees"
+            srs = s["srs"]
+        else: srs = 'EPSG:4326'
+        
+        tmp_dir = tempfile.mkdtemp() + "/" + sql_name
```
As above, should use `os.path.join`

ahinz commented on treemap/views.py:
```diff
@@ -1827,48 +1824,47 @@ def ogr_conversion(output_type, tree_sql, plot_sql, extension=None, name="trees"
+            command.append('-nlt')
+            command.append('POINT')
+        
+        done = subprocess.call(command)
```
From past experience, this call must be in an exception handler since the actual command contains the database name and password:

``` python
done = None
begin:
    done = subprocess.call(command)
except:
    raise Exception("ogr2ogr2 command failed (are the gdal binaries installed?)")
```

ahinz commented on treemap/views.py:
```diff
@@ -1827,48 +1824,47 @@ def ogr_conversion(output_type, tree_sql, plot_sql, extension=None, name="trees"
 
-    done = subprocess.call(command)
+        if done != 0: 
+            return render_to_json({'status':'error', 'command': command})
```
We must _not_ return the command without scrubbing the database password!

ahinz commented on treemap/views.py:
```diff
@@ -1935,12 +1931,15 @@ def advanced_search(request, format='json'):
+
     if format == "shp":
-        return ogr_conversion('ESRI Shapefile', tree_query, plot_query)
+        return ogr_conversion('ESRI Shapefile', [{'name':'trees', 'sql':tree_query}, {'name':'plots', 'sql':plot_query}])
```
Since the dictionaries are always the same it might look better to define variables:

``` python
trees = {'name': 'trees', 'sql': tree_query}
plots = {'name'; 'plots', 'sql': plot_query }
```

That way these lines like a bit less intimidating:

``` python
return ogr_conversion('ESRI Shapefile', [trees, plots])
```

ahinz commented:
This is a great update to the export code!

shalperin committed:
ogr_conversion: security, docs and x-platform

Better protect shell command in ogr_conversion.  Minor documentation changes.  Use os.path.join for paths.
